### PR TITLE
GeoTiffRasterSource Thread Safety

### PR DESCRIPTION
**Pull request by jbouffard**
# Overview

This PR makes reading from `GeoTiffRasterSource`s thread safe by synchronizing the `tiff` objects within these instances. That will ensure that the state of the data doesn't change as it's being accessed by different threads.

## Checklist

- [x] Add entry to CHANGELOG.md
- [x] Tests

Closes https://github.com/locationtech/geotrellis/issues/2954

### CODE REVIEW

File: vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceSpec.scala [hunk 1]
```diff
@@ -141,4 +143,48 @@ class GeoTiffRasterSourceSpec extends FunSpec with RasterMatchers with BetterRas
...
+    val rs = GeoTiffRasterSource(url)
+
+    it("GeoTiffRasterSource should be threadsafe") {
+      val res: List[Future[Option[Raster[MultibandTile]]]] = (0 to 100).toList.map { _ => Future {
+        rs.read()
+      } }
+
+      val fres: Future[List[Raster[MultibandTile]]] = Future.sequence(res).map(_.flatten)
+
+      Await.result(fres, Duration.Inf)
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceSpec.scala]: Instead of writing `Awaits` we can use `AsyncFunSpec` here.

Comment (pomadchin) [reply vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceSpec.scala]: You can also create a separate `Spec` if it is necessary.

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala [hunk 2]
```diff
@@ -57,7 +57,11 @@ case class GeoTiffRasterSource(
...
     }
-    if (it.hasNext) Some(convertRaster(it.next)) else None
+
+    // We want to use this tiff in different `RasterSource`s, so we
+    // need to lock it in order to garuntee the state of tiff when
+    // it's being accessed by a thread.
+    tiff.synchronized { if (it.hasNext) Some(convertRaster(it.next)) else None }
   }
 
   def read(bounds: GridBounds[Long], bands: Seq[Int]): Option[Raster[MultibandTile]] = {
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala]: The following test would fail: 

```scala
it("GeoTiffRasterSource readBounds() should be threadsafe") {
  val res: List[Future[Option[Raster[MultibandTile]]]] = (0 to 100).toList.map { _ => Future {
    source.read(source.gridBounds, 0 until source.bandCount)
  } }

  val fres: Future[List[Raster[MultibandTile]]] = Future.sequence(res).map(_.flatten)

  fres.map { rasters => assert(rasters.size == 101) }
}
```

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala [hunk 3]
```diff
@@ -72,18 +72,19 @@ case class GeoTiffReprojectRasterSource(
...
-    if (it.hasNext) Some(it.next) else None
+
+    read(bounds, bands)
   }
 
   def read(bounds: GridBounds[Long], bands: Seq[Int]): Option[Raster[MultibandTile]] = {
     val it = readBounds(List(bounds), bands)
-    if (it.hasNext) Some(it.next) else None
+
+    tiff.synchronized { if (it.hasNext) Some(it.next) else None }
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala]: should there be lock on the overview?

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala [hunk 4]
```diff
@@ -72,18 +72,19 @@ case class GeoTiffReprojectRasterSource(
...
   }
 
   override def readExtents(extents: Traversable[Extent], bands: Seq[Int]): Iterator[Raster[MultibandTile]] = {
     val bounds = extents.map(gridExtent.gridBoundsFor(_, clamp = true))
-    readBounds(bounds, bands)
+    tiff.synchronized { readBounds(bounds, bands) }
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala]: would it work?

File: vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class GeoTiffRasterSourceMultiThreadingSpec extends AsyncFunSpec {
+  val url = Resource.path("img/aspect-tiled.tif")
+  val source: GeoTiffRasterSource = new GeoTiffRasterSource(url)
+
+  implicit val ec = ExecutionContext.global
+
+  describe("Thread Safety") {
+    it("GeoTiffRasterSource should be threadsafe") {
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala]: The test coverage is not full, can you add separate tests for 
* `read()`
* `readBounds()`
* `readExtents()`

This will help you to understand do you need these locks around iterators or not.

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala (hunk 3 above)
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala]: Yeah, it looks like it does.

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala (hunk 4 above)
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala]: Nope, it's not necessary there.

File: vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+
+class GeoTiffRasterSourceMultiThreadingSpec extends AsyncFunSpec {
+  val url = Resource.path("img/aspect-tiled.tif")
+  val source: GeoTiffRasterSource = new GeoTiffRasterSource(url)
+
+  implicit val ec = ExecutionContext.global
+
+  /*
+  describe("GeoTiffRasterSource should be threadsafe") {
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala]: Is it an intention to comment out these tests?

Comment (jbouffard) [reply vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala]: Ah, no it wasn't

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala (hunk 2 above)
Comment (pomadchin) [reply vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala]: It is still a failing test: https://github.com/geotrellis/geotrellis-contrib/blob/d5c86d7664c58972538ea29bd34b5abba2363d8f/vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSource.scala#L69

File: vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      } }
+
+      val fres: Future[List[Raster[MultibandTile]]] = Future.sequence(res).map(_.flatten)
+
+      fres.map { rasters => assert(rasters.size == 101) }
+    }
+
+    it("readBounds") {
+      val res: List[Future[Iterator[Raster[MultibandTile]]]] = (0 to 100).toList.map { _ => Future {
+        source.readBounds(List(source.gridBounds), 0 until source.bandCount toSeq)
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala]: `source.read` should be there along with `readBounds` spec. Since it will cover a different call. 

The idea is to check all overloads and all unique calls.

File: vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      } }
+
+      val fres: Future[List[Raster[MultibandTile]]] = Future.sequence(res).map(_.flatten)
+
+      fres.map { rasters => assert(rasters.size == 101) }
+    }
+
+    it("readExtents") {
+      val res: List[Future[Iterator[Raster[MultibandTile]]]] = (0 to 100).toList.map { _ => Future {
+        source.readExtents(List(source.extent), 0 until source.bandCount toSeq)
```
Comment (pomadchin) [inline vlm/src/test/scala/geotrellis/contrib/vlm/geotiff/GeoTiffRasterSourceMultiThreadingSpec.scala]: I think forcing here `.toList` can be a nice stress test.

File: vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala [hunk 9]
```diff
@@ -98,8 +98,7 @@ case class GeoTiffReprojectRasterSource(
         extent = gridExtent.extentFor(targetPixelBounds, clamp = true),
         cols = targetPixelBounds.width.toInt,
         rows = targetPixelBounds.height.toInt)
-      val sourceExtent =
-        closestTiffOverview.synchronized { targetRasterExtent.extent.reprojectAsPolygon(backTransform, 0.001).envelope }
+      val sourceExtent = targetRasterExtent.extent.reprojectAsPolygon(backTransform, 0.001).envelope
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/geotiff/GeoTiffReprojectRasterSource.scala]: Why did you remove `synchronized`? It creates stacktraces related to extent reprojection, check out https://github.com/geotrellis/geotrellis-contrib/pull/206#pullrequestreview-260115791
